### PR TITLE
Prometheus: Prometheus frontend library for testing in the decoupled DS

### PR DESCRIPTION
**Pull request by bohandley**
WIP

**What is this feature?**
This is a frontend package for Prometheus components. We are in the research phase. 

**Why do we need this feature?**
We intend to use this library in testing decoupling Prometheus from core Grafana.

**Who is this feature for?**
The observability metrics squad.

**Special notes**
import it in your external plugin like this in dependencies in package.json
`"@grafana/prometheus": "file:../<path to grafana>/grafana/packages/grafana-prometheus"`

DoD

- [x] Test an import in core and external
- [x] Remove all vendor code
- [x] Write e2e test suite for Prometheus
  - [x] [add selectors to core Prometheus and write e2e tests for core Prometheus](https://github.com/grafana/grafana/pull/80333)
  - [x] add selectors to library 
  - [ ] write e2e tests for external plugin
- [ ] https://github.com/grafana/grafana/pull/81610
- [ ] Create exports
  - [x] config
  - [ ] query editor
  - [ ] annotations
  - [ ] variable query editor
  - [ ] data source
  - [ ] etc
- [x] Fix theme bug with light and dark to read from Context
- [ ] Use the shared/ folder that has been migrated to grafana/experimental

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-prometheus/.eslintrc [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "rules": {
+    // "no-restricted-imports": ["error", { "patterns": ["@grafana/runtime", "@grafana/ui", "@grafana/data", "@grafana/e2e/*"] }]
```
Comment (itsmylife) [inline packages/grafana-prometheus/.eslintrc:3]: I think we can just remove this part. We need to import from those places.

File: packages/grafana-prometheus/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# (2023-06-11)
+
+First public release
```
Comment (itsmylife) [inline packages/grafana-prometheus/CHANGELOG.md:3]: Is this created automatically?

File: packages/grafana-prometheus/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    "compilerOptions": {
+      "module": "commonjs",
+      "target": "es5",
+      "esModuleInterop": true
+    },
+    "transpileOnly": true
+  },
+  "include": ["../src", "./types"],
+  "extends": "@grafana/tsconfig"
+}
```
Comment (itsmylife) [inline packages/grafana-prometheus/tsconfig.json:26]: I think we can use @grafana/data package tsconfig.json

```json
{
  "compilerOptions": {
    "alwaysStrict": true,
    "declarationDir": "./compiled",
    "emitDeclarationOnly": true,
    "isolatedModules": true,
    "rootDirs": ["."]
  },
  "exclude": ["dist/**/*"],
  "extends": "@grafana/tsconfig",
  "include": ["src/**/*.ts*", "typings/jest", "../../public/app/types/jquery/*.ts", "../../public/app/types/*.d.ts"]
}
```

File: packages/grafana-prometheus/package.json [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    "LICENSE_APACHE2"
+  ],
+  "scripts": {
+    "build": "tsc -p ./tsconfig.build.json && rollup -c rollup.config.ts",
+    "clean": "rimraf ./dist ./compiled ./package.tgz",
+    "typecheck": "tsc --emitDeclarationOnly false --noEmit",
+    "prepack": "cp package.json package.json.bak && node ../../scripts/prepare-packagejson.js",
+    "postpack": "mv package.json.bak package.json"
+  },
+  "dependencies": {
```
Comment (itsmylife) [inline packages/grafana-prometheus/package.json:37]: This part needs a bit cleaning I think. For instance react is defined peer dependency and it is also in dependencies list.

**Comment by bohandley** [comment]
Coming back to this work in the new year after the hackathon, vacation and a few escalations.

New information gathered and here is a guide for the next steps: 

- [x] meet with Ivana to go over shared Visual Query Builder code used in Prom & Loki
  - [ ] [review PR by Ivana](https://github.com/grafana/grafana-experimental/pull/98) to refactor Prom/Loki code, removing it from core grafana and thus we can remove it from the Prometheus library
  - [ ] Test refactored shared code in the Prometheus data source, external data source and the library
- [x] Continue cleaning the code to remove
  - [ ] Fix all the any: Unexpected any. Specify a different type.
  - [ ] All styles should be written as objects: Styles should be written using objects.
  - [ ] Do not use any type assertions.
  - [ ] Use data-testid for E2E selectors instead of aria-label
- [ ] Refactor configuration to export only Prometheus specific configuration. 
  - [ ] Vendor specific configuration components should not exist in the library

**Comment by bohandley** [comment]
Bug: testing configuration imports in core grafana does not match light and dark styles. needs investigation

Screenshot 2024-01-04 at 3 18 03 PM

File: packages/grafana-prometheus/package.json [hunk 5]
```diff
@@ -0,0 +1,151 @@
+{
+  "author": "Grafana Labs",
+  "license": "Apache-2.0",
```
Comment (jackw) [inline packages/grafana-prometheus/package.json:3]: I've not read all the code but it seems this package is pretty much the entire plugin minus the module.ts and plugin.json? All code in the Grafana codebase (except the NPM packages that we publish publicly) are AGPL-3.0. Are we sure we want to change this to Apache-2.0? 🤔

File: packages/grafana-prometheus/package.json [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  "name": "@grafana/prometheus",
```
[the remark this answers is not in the document]
Comment (bohandley) [reply packages/grafana-prometheus/package.json:4]: So simple and elegant 💯

File: packages/grafana-prometheus/package.json [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    "prepack": "cp package.json package.json.bak && node ../../scripts/prepare-packagejson.js",
+    "postpack": "mv package.json.bak package.json"
+  },
+  "dependencies": {
+    "@emotion/css": "11.11.2",
+    "@floating-ui/react": "0.26.8",
+    "@grafana/data": "workspace:*",
+    "@grafana/experimental": "1.7.8",
+    "@grafana/faro-web-sdk": "^1.3.6",
+    "@grafana/runtime": "workspace:*",
```
Comment (jackw) [inline packages/grafana-prometheus/package.json:43]: Shouldn't this also be a peerDependency so it resolves to the singleton in Grafana core?

File: packages/grafana-prometheus/package.json (hunk 5 above)
Comment (bohandley) [reply packages/grafana-prometheus/package.json:3]: Good point! We had not decided on the license yet. We know that there are some finer details to iron out with the legal team for how the custom vendors will use this code. If people are creating their own Prometheus plugins, we want to be in alignment with Grafana for the license.

File: packages/grafana-prometheus/package.json (hunk 7 above)
Comment (bohandley) [reply packages/grafana-prometheus/package.json:43]: This makes sense to me.

**Comment by bohandley** [comment]
Closing in favor of https://github.com/grafana/grafana/pull/81641

**Pull request closed without merging**